engine/access/ingestion2/results_forest.go

```diff
@@ -0,0 +1,452 @@
+			runningCount := uint(0)
+
+			// Inspect all vertices in the tree forming from the latest persisted sealed result.
+			// Count all running pipelines, and start new ones up to the maxRunningCount.
+			// This will iterate over at most 2*maxRunningCount vertices.
+			f.visitAllAncestorsBFS(f.latestPersistedSealedResult.ResultID(), func(container *ExecutionResultContainer) bool {
+				state := container.pipeline.GetState()
+
+				switch state {
+				case pipeline.StateCanceled:
+					// TODO: free the pipeline's resources (not necessarily here)
+					return true
+
+				case pipeline.StateComplete:
+					if err := f.processCompleted(container.resultID); err != nil {
+						ctx.Throw(err)
+					}
+					return true
+
+				case pipeline.StatePending:
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+
+						core := pipeline.NewCore()
+						err := container.pipeline.Run(ctx, core)
+						if err != nil && !errors.Is(err, context.Canceled) {
+							ctx.Throw(fmt.Errorf("pipeline execution failed (result: %s): %w", container.resultID, err))
+						}
+					}()
+				}
+
+				runningCount++
+				return runningCount < f.maxRunningCount
+			})
+		}
+	}
+}
+
+// visitAllAncestorsBFS visits all ancestors of the given result ID in a breadth-first manner, and
+// calls the provided function on each ancestor.
+// If the function returns false, the traversal is stopped.
+func (f *ResultsForest) visitAllAncestorsBFS(resultID flow.Identifier, fn func(*ExecutionResultContainer) bool) {
+	queue := []flow.Identifier{resultID}
+	for len(queue) > 0 {
+		currentID := queue[0]
+		queue = queue[1:]
+
+		f.iterateChildren(currentID, func(child *ExecutionResultContainer) bool {
+			if !fn(child) {
+				return false
+			}
+			queue = append(queue, child.resultID)
+			return true
+		})
+	}
+}
+
+// processCompleted processes a completed pipeline and prunes the forest.
+// No errors are expected during normal operation.
+func (f *ResultsForest) processCompleted(resultID flow.Identifier) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	// first, ensure that the result ID is in the forest, otherwise the forest is in an inconsistent state
+	container, ok := f.getContainer(resultID)
+	if !ok {
+		return fmt.Errorf("state update from unknown result vertex %s", resultID)
+	}
+
+	// next, ensure that this result descends from the latest persisted sealed result, otherwise
+	// the forest is in an inconsistent state since persisting must be done sequentially
+	if container.resultID != f.latestPersistedSealedResult.ResultID() {
+		return fmt.Errorf("result %s does not match the latest persisted sealed result %s", container.resultID, f.latestPersistedSealedResult.ResultID())
+	}
+
+	// finally, prune the forest up to the latest persisted result's block view
+	latestPersistedView := container.blockHeader.View
+	err := f.pruneUpToView(latestPersistedView)
+	if err != nil {
+		return fmt.Errorf("failed to prune results forest (view: %d): %w", latestPersistedView, err)
+	}
+
```

Comment:
can we do this right after calling `Run` on the pipeline? 👇 
https://github.com/onflow/flow-go/blob/391632a0525f8afd694d3aca312a3ae18586725a/engine/access/ingestion2/results_forest.go#L131-L135

Reply:
yea, that's a good point